### PR TITLE
icu-port: build common, i18n, stubdata

### PR DESCRIPTION
**Pull request by sps-gold**
Changes are made to build following icu libraries:
'libicu_common.a', 
'libicu_stubdata.a', 
'libicu_i18n.a'

This will allow to load the icudt file. Likely resolves - https://github.com/emscripten-core/emscripten/issues/10129

This change depends on https://github.com/emscripten-core/emscripten/pull/13579 

This change also modifies install headers code, it was required to allow placing of header files for different libraries of one component into one folder. Like in this case I am placing all required headers under 'unicode' folder.

### CODE REVIEW

**Comment by sps-gold** [comment]
fyip @kripken @sbc100

File: tools/ports/icu.py [hunk 1]
```diff
@@ -7,9 +7,13 @@
 import os
 import shutil
 
-TAG = 'release-62-1'
-VERSION = '62_1'
-HASH = 'd3fa42da9aa9c2fc749fff4a31a9e57e826903681d9f4e5b4474649bf3efe271fec10f214a027d542123b85ad3f6fcfc9b6208ad3f8e4c24fe4a0cbab4024e2d'
+TAG = 'release-68-2'
+VERSION = '68_2'
+HASH = '12c3db5966c234c94e7918fb8acc8bd0838edc36a620f3faa788e7ff27b06f1aa431eb117401026e3963622b9323212f444b735d5c9dd3d0b82d772a4834b993'
```
Comment (sbc100) [inline tools/ports/icu.py:12]: Does this change require an upgrade?   It might be nice the land the upgrade as a separate PR in case there is any fallout it would make it easier to determine the cause.

File: tools/ports/icu.py [hunk 2]
```diff
@@ -20,24 +24,74 @@ def get(ports, settings, shared):
...
+  def get_downloaded_icu4c_path():
+    return os.path.join(ports.get_dir(), 'icu', 'icu')
+
+  def get_icu_build_path():
+    return os.path.join(ports.get_build_dir(), 'icu')
 
-    source_path = os.path.join(ports.get_dir(), 'icu', 'icu')
-    dest_path = os.path.join(ports.get_build_dir(), 'icu')
+  def get_icu_source_path():
+    return os.path.join(get_icu_build_path(), 'source')
```
Comment (sbc100) [inline tools/ports/icu.py:34]: Could these three function not just be local variables?  e.g.:

```
downloaded_icu4c_path = os.path.join(ports.get_dir(), 'icu', 'icu')
```

It seems like making them function make things more verbose and less efficient (due to re-computation).

File: tools/system_libs.py [hunk 3]
```diff
@@ -1565,13 +1565,23 @@ def get_include_dir():
...
   @staticmethod
-  def install_header_dir(src_dir, target=None):
+  def install_header_dir(src_dir, target=None, merge=False):
     if not target:
       target = os.path.basename(src_dir)
     dest = os.path.join(Ports.get_include_dir(), target)
-    shared.try_delete(dest)
-    logger.debug('installing headers: ' + dest)
-    shutil.copytree(src_dir, dest)
+    if not merge:
```
Comment (sbc100) [inline tools/system_libs.py:1572]: Perhaps flip the conditions to the condition is in the affirmative (`if merge:`...`)

Comment (sbc100) [reply tools/system_libs.py:1572]: Isn't the the main different between `install_header_dir` and `install_headers` that with the former the entire directory is owned by the port (and therefore deleted before install), whereas the with the later we have to merge in with other headers (so we can't just delete it).

I wonder if just using `install_headers` would work instead of adding this extra arg?

File: tools/ports/icu.py [hunk 4]
```diff
@@ -20,24 +24,74 @@ def get(ports, settings, shared):
...
+        # generate static
+        '-DU_STATIC_IMPLEMENTATION',
+        # CXXFLAGS
+        '-std=c++11'
+    ]
+    ports.build_port(lib_src, lib_output, other_includes, build_flags + additional_build_flags)
+
+  # creator for libicu_common
+  def create_libicu_common(lib_output):
+    prepare_build()
```
Comment (sbc100) [inline tools/ports/icu.py:63]: I guess the other two build function rely on the fact that this one always run first?  (They assume prepare_build has already been called?)

File: tools/ports/icu.py [hunk 5]
```diff
@@ -20,24 +24,74 @@ def get(ports, settings, shared):
...
+        '-DU_USING_ICU_NAMESPACE=0',
+        # make explicit inclusion of utf header: ref utf.h
+        '-DU_NO_DEFAULT_INCLUDE_UTF_HEADERS=1',
+        # mark UnicodeString constructors explicit : ref unistr.h
+        '-DUNISTR_FROM_CHAR_EXPLICIT=explicit',
+        '-DUNISTR_FROM_STRING_EXPLICIT=explicit',
+        # generate static
+        '-DU_STATIC_IMPLEMENTATION',
+        # CXXFLAGS
+        '-std=c++11'
```
Comment (sbc100) [inline tools/ports/icu.py:57]: It seems odd than non of these were needed before.  Do you know why?  Is it just because we are building so much more of the library?

**Review by sbc100 — COMMENTED** [review]
Thanks for working on this!

I do wonder if we could do this more incrementally.  For example is it reasonable to do the library upgrade first a separate PR?

**Comment by sps-gold** [comment]
hi @sbc100, I believe I have incorporated all your review comments. please have a relook and merge if it looks good

**Review by sbc100 — APPROVED** [review]
